### ISSUE TITLE
[Refactoring] conceptual limit/bug of addMethod refactoring

### ISSUE THREAD

**Issue by Ducasse** [guillep/pharo-core#17235]
Imagine the following situation

```
A >> x ^ 33
   B is empty
      C >> x ^ super x + 10 
```

Now we add a simple method x in B. 
So we are changing behavior if its implementation is different from the one of A!
Bingo

Even funnier

```
A >> x ^ 1
    B >> x ^ super x + 12  
          C is empty
              D >> x ^ super x + 10 
```
Now we add a simple method to C `C >> x ^ super x + 12`.
We are changing the behavior of `D >> x`.

**Comment by Ducasse** [comment]
Maybe we are the first ones to realise that Add Method is complex!

**Comment by Ducasse** [comment]
Imagine now 

```
A >> x ^ 12
    B >> isEmpty
      C>> foo ^ self x
```

Now we add in B x ^ 42
We got a behavior change.

**Comment by balsa-sarenac** [comment]
We should check this for extract method, since add method has this check

**Comment by AlexisCnockaert** [comment]
The `ComposedExtractMethod` refactoring has this check too in `preconditionNoOverrides` so I think the issue can be closed

**Comment by balsa-sarenac** [comment]
I've checked `ExtractMethod` and was able to reproduce but only when option is not set.
So this was issue with regular extract with specific option configuration.
The ExtractMethod has option for should override methods, but it's hidden in `getNewMethodName` as option check.
We should pull it up to be regular behavior-preserving precondition.

**Comment by AlexisCnockaert** [comment]
Ok sorry I didn't pay attention, well the thing is `getNewMethodName` is called in `privateTransform`, so if we create a behaviour-preserving precondition for this, the refactoring won't know the new selector to check in the execution flow

**Comment by balsa-sarenac** [comment]
Yes, it's a tricky issue. The path forward is to move this logic for getting the new selector to the driver, but since ExtractMethod is so complex, it's going very slowly...
We can do it in a couple of iterations
But yes, extract method is overdue for a huge restructuring. I don't think we should do it now, but maybe over the summer. Maybe @jainoshika might be interested in doing it?

**Comment by Ducasse** [comment]
indeed.

**Comment by jainoshika** [comment]
sure @balsa-sarenac, i would love to work on this.
